src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala

```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+    def visitAnnotationss(annotss: Array[java.util.List[AnnotationNode]]) = if (annotss != null) annotss foreach visitAnnotations
+
+    def visitHandle(handle: Handle): Unit = {
+      visitInternalNameOrArrayReference(handle.getOwner)
+      visitDescriptor(handle.getDesc)
+    }
+  }
+
+  abstract class GenericSignatureVisitor {
+    def visitInternalName(internalName: InternalName): Unit
+
+    def raiseError(msg: String, sig: String, e: Option[Throwable] = None): Unit
+
+    def visitClassSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.classSignature() }
+    }
+
+    def visitMethodSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.methodSignature() }
+    }
+
+    def visitFieldSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.fieldSignature() }
+    }
+
+    private final class Parser(sig: String) {
+      private var index = 0
+      private val end = sig.length
+
+      private val Aborted: Throwable = new NoStackTrace { }
+      private def abort(): Nothing = throw Aborted
+
+      def safely(f: => Unit): Unit = try f catch {
+        case Aborted =>
+        case NonFatal(e) => raiseError(s"Exception thrown during signature parsing", sig, Some(e))
+      }
+
+      private def current = {
+        if (index >= end) {
+          raiseError(s"Out of bounds, $index >= $end", sig)
+          abort() // Don't continue, even if `notifyInvalidSignature` returns
+        }
+        sig.charAt(index)
+      }
+
+      private def accept(c: Char): Unit = {
+        if (current != c) {
+          raiseError(s"Expected $c at $index, found $current", sig)
+          abort()
+        }
+        index += 1
+      }
+
+      private def skip(): Unit = { index += 1 }
+      private def getCurrentAndSkip(): Char = { val c = current; skip(); c }
+
+      private def skipUntil(isDelimiter: Char => Boolean): Unit = {
+        while (!isDelimiter(current)) { index += 1 }
+      }
+
+      private def appendUntil(builder: java.lang.StringBuilder, isDelimiter: Char => Boolean): Unit = {
+        val start = index
+        skipUntil(isDelimiter)
+        builder.append(sig, start, index)
+      }
+
+      def isBaseType(c: Char): Boolean = c match {
+        case 'B' | 'C' | 'D' | 'F' | 'I' | 'J' | 'S' | 'Z' => true
+        case _ => false
+      }
+
+      private val isClassNameEnd = (c: Char) => c == '<' || c == '.' || c == ';'
+
+      private def typeArguments(): Unit = if (current == '<') {
+        skip()
+        while (current != '>') current match {
+          case '*' | '+' | '-' =>
+            skip()
+          case _ =>
+            referenceTypeSignature()
+        }
+        accept('>')
+      }
+
+      @tailrec private def referenceTypeSignature(): Unit = getCurrentAndSkip() match {
+        case 'L' =>
+          val names = new java.lang.StringBuilder()
```

Comment:
I did a quick benchmark (https://github.com/lrytz/compiler-benchmark/commits/nestedClassCollector) for the various ideas (revisions here https://github.com/lrytz/scala/commits/t10180-wip)

```
* e5ea3abfa3 - (origin/t10180) Parse signatures when collecting referenced nested classes (8 days ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-e5ea3ab -DscalaRef=e5ea3ab -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  4.902 ± 0.104  ops/s

* a34aa97968 - unnecessary ListBuffer (2 hours ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-a34aa97 -DscalaRef=a34aa97 -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  4.876 ± 0.116  ops/s

* 2132e98f13 - charBooleanFun (2 hours ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-2132e98 -DscalaRef=2132e98 -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  5.146 ± 0.104  ops/s

* c8f7eb73b5 - initial size (2 hours ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-c8f7eb7 -DscalaRef=c8f7eb7 -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  5.148 ± 0.101  ops/s

* 4b4f863c46 - (HEAD -> t10180-wip, origin/t10180-wip) single builder (2 hours ago) <Lukas Rytz>
# VM options: -DscalaVersion=2.12.3-bin-4b4f863 -DscalaRef=4b4f863 -Dsbt.launcher=/Users/luc/.sbt/launchers/0.13.15/sbt-launch.jar
NestedClassCollector.rtJar  thrpt   20  5.115 ± 0.129  ops/s
```

So the `CharBooleanFunction` helped, but then setting an initial size for the buffer here or using a single buffer didn't make a difference.